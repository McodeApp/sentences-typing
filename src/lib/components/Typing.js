import React, {Component} from 'react';
import PropTypes from 'prop-types';

class Typing extends Component {
    state = {
        currentIndex: 1,
        firstRow: '',
        secondRow: '',
        firstRowFinish: false,
        typeSpeed: 100,
    };

    componentDidMount() {
        this.type();
    }

    type = () => {
        const {currentIndex, firstRowFinish, secondRow, typeSpeed, firstRow} = this.state;
        const {sentences, infinity} = this.props;
        let options = {};

        if (sentences.length === 1 && firstRowFinish) {
            return;
        }

        if (currentIndex === sentences.length - 1 && secondRow === sentences[currentIndex]) {
            if (infinity) {
                options = {
                    currentIndex: 1,
                    secondRow: '',
                }
            } else {
                return;
            }
        }

        if (!firstRowFinish && sentences.length) {
            const [first] = sentences;
            const firstRowText = first.substring(0, firstRow.length + 1);
            options ={
                firstRow: firstRowText,
                firstRowFinish: first.length === firstRow.length,
            };
        } else if (currentIndex < sentences.length) {
            if (secondRow.length !== sentences[currentIndex].length) {
                const secondRowText = sentences[currentIndex].substring(0, secondRow.length + 1);
                options = {
                    secondRow: secondRowText,
                    currentIndex: currentIndex,
                };
            } else {
                const index = currentIndex < sentences.length -1 ? currentIndex + 1 : 1;
                options = {
                    secondRow: '',
                    currentIndex: index,
                };
            }
        }

        this.setState(options);
        setTimeout(this.type, typeSpeed);
    };

    render() {
        const {firstRow, secondRow} = this.state;
        return (
            <div>
                <h1>{firstRow}</h1>
                {secondRow && <h1>{secondRow}</h1>}
            </div>
        );
    }
}

Typing.propTypes = {
    infinity: PropTypes.bool,
    sentences: PropTypes.arrayOf(PropTypes.string).isRequired,
};

Typing.defaultProps = {
    infinity: false,
};

export default Typing;
