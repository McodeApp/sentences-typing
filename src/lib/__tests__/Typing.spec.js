import React from 'react';
import Typing from '../components/Typing';
import renderer from 'react-test-renderer'

jest.useFakeTimers();

const sentences = [
    "1234567890",
    "qwerty",
    "dfksjn32",
];

const runTimersToTime = (sentence, clearNextLine = false) => {
    let time = sentence.length * typeSpeed;
    time = clearNextLine ? time + typeSpeed : time;
    jest.runTimersToTime(time);
};

const [first, second, three] = sentences;
const typeSpeed = 100;

describe('Typing sentence', () => {
    describe('One sentence', () => {
        it("renders properly one sentence", () => {
            const tree = renderer.create(<Typing sentences={[first]}/>);

            runTimersToTime(first);
            expect(tree.toJSON()).toMatchSnapshot();
        })
    });

    describe('Two sentences', () => {
        it("renders properly two sentences", () => {
            const tree = renderer.create(<Typing sentences={[first, second]}/>);

            runTimersToTime(first);
            expect(tree.toJSON()).toMatchSnapshot();
            runTimersToTime(second);
            expect(tree.toJSON()).toMatchSnapshot();
        })
    });

    describe('Three sentences', () => {
        it("renders properly three sentences", () => {
            const tree = renderer.create(<Typing sentences={[first, second, three]} infinity/>);

            runTimersToTime(first);
            expect(tree.toJSON()).toMatchSnapshot();
            runTimersToTime(second);
            expect(tree.toJSON()).toMatchSnapshot();
            runTimersToTime(three, true);
            expect(tree.toJSON()).toMatchSnapshot();
        })
    });

    describe('Three sentences - infinity mode', () => {
        it("renders properly three sentences", () => {
            const tree = renderer.create(<Typing sentences={[first, second, three]} infinity/>);

            runTimersToTime(first);
            expect(tree.toJSON()).toMatchSnapshot();
            runTimersToTime(second);
            expect(tree.toJSON()).toMatchSnapshot();
            runTimersToTime(three, true);
            expect(tree.toJSON()).toMatchSnapshot();
            runTimersToTime(second, true);
            expect(tree.toJSON()).toMatchSnapshot();
            runTimersToTime(three, true);
            expect(tree.toJSON()).toMatchSnapshot();
        })
    });
});