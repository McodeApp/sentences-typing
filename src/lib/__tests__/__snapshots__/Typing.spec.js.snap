// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typing sentence One sentence renders properly one sentence 1`] = `
<div>
  <h1>
    1234567890
  </h1>
  
</div>
`;

exports[`Typing sentence Three sentences - infinity mode renders properly three sentences 1`] = `
<div>
  <h1>
    1234567890
  </h1>
  
</div>
`;

exports[`Typing sentence Three sentences - infinity mode renders properly three sentences 2`] = `
<div>
  <h1>
    1234567890
  </h1>
  <h1>
    qwerty
  </h1>
</div>
`;

exports[`Typing sentence Three sentences - infinity mode renders properly three sentences 3`] = `
<div>
  <h1>
    1234567890
  </h1>
  <h1>
    dfksjn32
  </h1>
</div>
`;

exports[`Typing sentence Three sentences - infinity mode renders properly three sentences 4`] = `
<div>
  <h1>
    1234567890
  </h1>
  <h1>
    qwerty
  </h1>
</div>
`;

exports[`Typing sentence Three sentences - infinity mode renders properly three sentences 5`] = `
<div>
  <h1>
    1234567890
  </h1>
  <h1>
    dfksjn32
  </h1>
</div>
`;

exports[`Typing sentence Three sentences renders properly three sentences 1`] = `
<div>
  <h1>
    1234567890
  </h1>
  
</div>
`;

exports[`Typing sentence Three sentences renders properly three sentences 2`] = `
<div>
  <h1>
    1234567890
  </h1>
  <h1>
    qwerty
  </h1>
</div>
`;

exports[`Typing sentence Three sentences renders properly three sentences 3`] = `
<div>
  <h1>
    1234567890
  </h1>
  <h1>
    dfksjn32
  </h1>
</div>
`;

exports[`Typing sentence Two sentences renders properly two sentences 1`] = `
<div>
  <h1>
    1234567890
  </h1>
  
</div>
`;

exports[`Typing sentence Two sentences renders properly two sentences 2`] = `
<div>
  <h1>
    1234567890
  </h1>
  <h1>
    qwerty
  </h1>
</div>
`;
